File: Anul II/Medii de proiectare si programare/Laborator/lab05/Ubb.BikeContest.Repository/src/RaceEntryDbRepository.cs
using System.Collections;
using System.Data;
using log4net;
using Ubb.BikeContest.Model;

namespace Ubb.BikeContest.Repository;

public class RaceEntryDbRepository : IRaceEntryRepository
{
    private static readonly ILog Log = LogManager.GetLogger("RaceEntryDbRepository");
    private readonly IDictionary<string, string?> _props;
    private readonly IRaceRepository _raceRepository;
    private readonly IParticipantRepository _participantRepository;

    public RaceEntryDbRepository(IDictionary<string, string?> props, 
        IRaceRepository raceRepository, IParticipantRepository participantRepository)
    {
        Log.Info("Creating RaceEntryDbRepository ");
        _props = props;
        _raceRepository = raceRepository;
        _participantRepository = participantRepository;
    }

    public RaceEntry FindById(long id)
    {
        Log.InfoFormat("Entering Get with value {0}", id);
        var connection = DbUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,participantID,raceID FROM race_entries WHERE id=@id";
            var paramId = command.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            command.Parameters.Add(paramId);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    RaceEntry raceEntry = Extract(dataReader);
                    Log.InfoFormat("Exiting Get with value {0}", raceEntry);
                    return raceEntry;
                }
            }
        }
        Log.InfoFormat("Exiting Get with value {0}", null);
        return null;
    }

    public IEnumerable<RaceEntry> FindAll()
    {
        Log.InfoFormat("Entering Read");
        var connection = DbUtils.GetConnection(_props);
        IList<RaceEntry> raceEntries = new List<RaceEntry>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,participantID,raceID FROM race_entries";
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    RaceEntry raceEntry = Extract(dataReader);
                    raceEntries.Add(raceEntry);
                }
            }
        }
        Log.InfoFormat("Exiting Read with value {0}", raceEntries);
        return raceEntries;
    }

    public void Save(RaceEntry newEntity)
    {
        Log.InfoFormat("Entering Add with value {0}", newEntity);
        var connection = DbUtils.GetConnection(_props);

        using var command = connection.CreateCommand();
        command.CommandText = 
            "INSERT INTO race_entries(participantID, raceID) " +
            "VALUES(@participantID, @raceID);";
        var participantId = command.CreateParameter();
        participantId.ParameterName = "@participantID";
        participantId.Value = newEntity.Participant.Id;
        command.Parameters.Add(participantId);

        var raceId = command.CreateParameter();
        raceId.ParameterName = "@raceID";
        raceId.Value = newEntity.Race.Id;
        command.Parameters.Add(raceId);

        var result = command.ExecuteNonQuery();
        Log.InfoFormat("Added {0} entities", result);
    }

    public void Delete(long id)
    {
        Log.InfoFormat("Entering Delete with value {0}", id);
        var connection = DbUtils.GetConnection(_props);
        using var command = connection.CreateCommand();
        command.CommandText = "DELETE FROM race_entries WHERE id=@id";
        var paramId = command.CreateParameter();
        paramId.ParameterName = "@id";
        paramId.Value = id;
        command.Parameters.Add(paramId);
        var result = command.ExecuteNonQuery();
        Log.InfoFormat("Deleted {0} entities", result);
    }

    public void Update(RaceEntry updatedEntity)
    {
        throw new NotImplementedException();
    }

    public IEnumerable<RaceEntry> GetEntriesByRace(long raceId)
    {
        Log.InfoFormat("Entering GetEntriesByRace");
        var connection = DbUtils.GetConnection(_props);
        IList<RaceEntry> raceEntries = new List<RaceEntry>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,participantID,raceID FROM race_entries WHERE raceID=@raceID";
            var paramRaceId = command.CreateParameter();
            paramRaceId.ParameterName = "@raceID";
            paramRaceId.Value = raceId;
            command.Parameters.Add(paramRaceId);
            
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    RaceEntry raceEntry = Extract(dataReader);
                    raceEntries.Add(raceEntry);
                }
            }
        }
        Log.InfoFormat("Exiting GetEntriesByRace with value {0}", raceEntries);
        return raceEntries;
    }

    public IEnumerable<Race> GetRacesWhereNotRegisteredAndEngineCapacity(long participantId, int engineCapacity)
    {
        Log.InfoFormat("Entering GetRacesWhereNotRegisteredAndEngineCapacity with {0}, {1}",
            participantId, engineCapacity);
        var connection = DbUtils.GetConnection(_props);
        IList<Race> result = new List<Race>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT * FROM races WHERE id NOT IN " +
                                  "(SELECT raceID from race_entries WHERE participantID=@participantId)" +
                                  " AND engineCc=@engineCc";
            var paramParticipantId = command.CreateParameter();
            paramParticipantId.ParameterName = "@participantId";
            paramParticipantId.Value = participantId;
            command.Parameters.Add(paramParticipantId);
            
            var paramEngine = command.CreateParameter();
            paramEngine.ParameterName = "@engineCc";
            paramEngine.Value = engineCapacity;
            command.Parameters.Add(paramEngine);
            
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Race race = ExtractRace(dataReader);
                    result.Add(race);
                }
            }
        }
        Log.InfoFormat("Exiting GetRacesWhereNotRegisteredAndEngineCapacity with value {0}", result);
        return result;
    }

    private RaceEntry Extract(IDataReader dataReader)
    {
        var id = dataReader.GetInt64(0);
        var participantId = dataReader.GetInt64(1);
        var raceId = dataReader.GetInt64(2);

        var participant = _participantRepository.FindById(participantId);
        var race = _raceRepository.FindById(raceId);
        var raceEntry = new RaceEntry(participant, race);
        return raceEntry;
    }
    
    private Race ExtractRace(IDataReader dataReader)
    {
        var id = dataReader.GetInt64(0);
        var name = dataReader.GetString(1);
        var engineCapacity = dataReader.GetInt32(2);

        var race = new Race(name, engineCapacity)
        {
            Id = id
        };
        return race;
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab08/Java/persistence/src/main/java/com/ubb/repository/jdbc/RaceEntryDBRepository.java
package com.ubb.repository.jdbc;

import com.ubb.model.Participant;
import com.ubb.model.Race;
import com.ubb.model.RaceEntry;
import com.ubb.repository.IParticipantRepository;
import com.ubb.repository.IRaceEntryRepository;
import com.ubb.repository.IRaceRepository;
import com.ubb.utils.JDBCUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import java.util.Properties;

public class RaceEntryDBRepository implements IRaceEntryRepository {
    private final JDBCUtils jdbcUtils;
    private final IRaceRepository raceRepository;
    private final IParticipantRepository participantRepository;
    private static final Logger logger = LogManager.getLogger();

    public RaceEntryDBRepository(
            final Properties properties, final IRaceRepository raceRepository,
            final IParticipantRepository participantRepository) {
        logger.info("Initializing RaceEntryDBRepository with properties: {} ", properties);
        jdbcUtils = new JDBCUtils(properties);
        this.raceRepository = raceRepository;
        this.participantRepository = participantRepository;
    }
    @Override
    public Collection<RaceEntry> getEntriesByRace(Long raceID) {
        logger.traceEntry("getEntriesByRace with task of id={} ", raceID);
        Connection connection = jdbcUtils.getConnection();
        Collection<RaceEntry> raceEntries = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM race_entries WHERE raceID=?")) {
            preparedStatement.setLong(1, raceID);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    RaceEntry raceEntry = extract(resultSet);
                    raceEntries.add(raceEntry);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit(raceEntries);
        return raceEntries;
    }

    @Override
    public void deleteByIds(Long participantID, Long raceID) {
        logger.traceEntry("deleting task of participantID={}, raceID={} ", participantID, raceID);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "DELETE FROM race_entries WHERE participantID=? AND raceID=?")) {
            preparedStatement.setLong(1, participantID);
            preparedStatement.setLong(2, raceID);
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Deleted {} instances", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public Collection<Race> getRacesWhereNotRegisteredAndEngineCapacity(Long participantID, Integer engineCc) {
        logger.traceEntry("getRacesWhereNotRegistered with tasks {}, {} ", participantID, engineCc);
        Connection connection = jdbcUtils.getConnection();
        Collection<Race> raceEntries = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM races WHERE id NOT IN " +
                        "(SELECT raceID from race_entries WHERE participantID=?)" +
                        " AND engineCc=?")) {
            preparedStatement.setLong(1, participantID);
            preparedStatement.setInt(2, engineCc);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Race race = extractRace(resultSet);
                    raceEntries.add(race);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit(raceEntries);
        return raceEntries;
    }

    @Override
    public Optional<RaceEntry> findById(Long id) {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "select * from race_entries where id=?")) {
            preparedStatement.setLong(1, id);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    RaceEntry raceEntry = extract(resultSet);
                    logger.traceExit(raceEntry);
                    return Optional.of(raceEntry);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }

    @Override
    public Collection<RaceEntry> findAll() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        Collection<RaceEntry> raceEntries = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM race_entries")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    RaceEntry raceEntry = extract(resultSet);
                    raceEntries.add(raceEntry);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit(raceEntries);
        return raceEntries;
    }

    @Override
    public void save(RaceEntry newEntity) {
        logger.traceEntry("updating with task {} ", newEntity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "INSERT INTO race_entries (participantID, raceID) values (?,?)")) {
            preparedStatement.setLong(1, newEntity.getParticipant().getId());
            preparedStatement.setLong(2, newEntity.getRace().getId());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Updated {} instances", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public void delete(Long id) {
        logger.traceEntry("deleting task of id={} ", id);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "DELETE FROM race_entries where id=?")) {
            preparedStatement.setLong(1, id);
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Deleted {} instances", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public void update(RaceEntry updatedEntity) {
    }

    private RaceEntry extract(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        Long participantID = resultSet.getLong("participantID");
        Long raceID = resultSet.getLong("raceID");

        final Optional<Race> race = raceRepository.findById(raceID);
        final Optional<Participant> participant = participantRepository.findById(participantID);
        if (participant.isPresent() && race.isPresent()) {
            RaceEntry raceEntry = new RaceEntry(participant.get(), race.get());
            raceEntry.setId(id);
            return raceEntry;
        } else {
            throw new SQLException("Database error: Unable to reference participant and/or race for Race Entry");
        }
    }

    private Race extractRace(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        String name = resultSet.getString("name");
        Integer engineCapacity = resultSet.getInt("engineCc");

        Race race = new Race(name, engineCapacity);
        race.setId(id);
        return race;
    }
}
